backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt

```diff
@@ -0,0 +1,191 @@
+import org.jetbrains.kotlin.backend.konan.ir.KonanSymbols
+import org.jetbrains.kotlin.backend.konan.ir.superClasses
+import org.jetbrains.kotlin.ir.builders.IrBuilderWithScope
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.builders.irGetObject
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrFunction
+import org.jetbrains.kotlin.ir.declarations.isPropertyAccessor
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.classOrNull
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.util.*
+
+/**
+ * Check given function is a getter or setter
+ * for `value` property of CEnumVar subclass.
+ */
+private fun isEnumVarValueAccessor(function: IrFunction, symbols: KonanSymbols): Boolean {
+    val parent = function.parent as? IrClass ?: return false
+    return if (symbols.interopCEnumVar in parent.superClasses && function.isPropertyAccessor) {
+        function.propertyIfAccessor.nameForIrSerialization.asString() == "value"
+    } else {
+        false
+    }
+}
+
+private class InteropCallContext(
+        val symbols: KonanSymbols,
+        val builder: IrBuilderWithScope
+)
+
+private inline fun <T> generateInteropCall(
+        symbols: KonanSymbols,
+        builder: IrBuilderWithScope,
+        block: InteropCallContext.() -> T
+) = InteropCallContext(symbols, builder).block()
+
+/**
+ * Search for memory read/write function in [kotlinx.cinterop.nativeMemUtils] of a given [valueType].
+ */
+private fun InteropCallContext.findMemoryAccessFunction(isRead: Boolean, valueType: IrType): IrFunction {
+    val requiredType = if (isRead) {
+        "INTEROP_READ_PRIMITIVE"
+    } else {
+        "INTEROP_WRITE_PRIMITIVE"
+    }
+    val nativeMemUtilsClass = symbols.nativeMemUtils.owner
+    return nativeMemUtilsClass.functions.filter {
+        val annotationArgument = it.annotations
+                .findAnnotation(RuntimeNames.typedIntrinsicAnnotation)
+                ?.getAnnotationStringValue()
+        annotationArgument == requiredType
+    }.first {
+        if (isRead) {
+            it.returnType.classOrNull == valueType.classOrNull
+        } else {
+            it.valueParameters.last().type.classOrNull == valueType.classOrNull
+        }
+    }
+}
+
+private fun InteropCallContext.readPrimitiveFromMemory(
+        nativePointed: IrExpression,
+        returnType: IrType
+): IrExpression  {
+    val memoryValueType = determineInMemoryType(returnType)
+    val memReadFn = findMemoryAccessFunction(isRead = true, valueType = memoryValueType)
+    val memRead = builder.irCall(memReadFn).also { memRead ->
+        memRead.dispatchReceiver = builder.irGetObject(symbols.nativeMemUtils)
+        memRead.putValueArgument(0, builder.irCall(symbols.interopInterpretNullablePointed).also {
```

Comment:
Why nullable?

Reply:
Because `interpretNullablePointed` is an intrinsic.